data_pull.py

```diff
@@ -53,33 +81,158 @@ def pull_data(self, search_criteria: str, delete_csv: bool = False):
+        multi-family, mobile, co-op, and other. If selecting more than one home type, initialize param as a list
+        :param delete_csv: When set to True, deletes the CSV file that is downloaded as part of the data pulling process
+        :return: Pandas DataFrame containing real estate data from Redfin for the desired city & state
+        """
+        # Identify search bar on home page for entering a real estate search location and search for input city/state
+        # Finds first search bar on the home page, which is the one we're looking for
+        driver.find_element(By.CLASS_NAME, 'search-input-box').send_keys(f'{search_criteria}' + Keys.ENTER)
+        time.sleep(1.5)
+
+        # Click 'All Filters'; Set to show 'For Sale' listings only by default
+        driver.find_element(By.XPATH, '//*[@id="sidepane-header"]/div/div[1]/form/div[5]/div').click()
+        time.sleep(1.5)
+
+        # Select desired 'Home Type'
+        home_type_header = driver.find_element(By.XPATH, '//*[@id="filterContent"]/div/div[5]/div[1]/div/span')
+        driver.execute_script("return arguments[0].scrollIntoView();", home_type_header)
+        time.sleep(1.5)
+
+        if type(home_types) == str:     # If user selects single home type (string), convert it to a list
+            home_types = [home_types]
+        home_type_select(home_types=home_types, driver=driver)
+        time.sleep(1.5)
+
+        # Listing status set to both 'Coming Soon' and 'Active' by default; Uncheck 'Coming Soon'
+        driver.find_element(By.XPATH, '//*[@id="filterContent"]/div/div[6]/div[1]/div/div[2]/span[1]/label/span[1]').click()
+        time.sleep(1.5)
+
+        # Set to only include listings from the last 7 days (reduces # of homes to download; need to keep < 350)
+        select = Select(driver.find_element(By.XPATH, '//*[@id="filterContent"]/div/div[6]/div[2]/div[2]/span/span/select'))
+        select.select_by_visible_text('Less than 7 days')
+        time.sleep(1.5)
+
+        # Turn off 'Foreclosures' listing type
+        listing_type_header = driver.find_element(By.XPATH, '//*[@id="filterContent"]/div/div[10]/div[1]/div/span')
+        driver.execute_script('return arguments[0].scrollIntoView()', listing_type_header)
+        time.sleep(1.5)
+
+        driver.find_element(By.XPATH, '//*[@id="filterContent"]/div/div[10]/div[2]/div/div[1]/div[2]/span/label/span[1]').click()
+        time.sleep(1.5)
+
+        # Close 'All Filters' menu
+        driver.find_element(By.XPATH, '//*[@id="right-container"]/div[6]/div/aside/header/button').click()
+        time.sleep(1.5)
+
+        # Click the (Download All) link at the bottom of results page to download a CSV with data from all real estate
+        # listings for every page of the search results
+        old_num_files = len(os.listdir(download_dir))  # Check number of files in download path prior to download
+        file_type = '*.csv'
+        driver.find_element(By.ID, 'download-and-save').click()
```

Comment:
Here's what I'd do from an object-oriented programming (OOP) perspective. 

I'd have a generic SeleniumWebDriverBuilder class which simply contains a method to construct your Selenium web driver.

Then I'd have a RedfinDataExtractor class that in its init method calls SeleniumWebDriverBuilder().buildDriver() which returns a driver which is stored as a variable. Still within the init, you could do what's done in lines 87-97. What this does is makes the code more flexible for use with other web pages you may want to use different settings for in the future and may need a different web driver.

The RedfinDataExtractor could then have methods like get_listing_data(), get_sales_data(). What may be cool is have those be handled by different classes too. So you'd have a RedfinListingDataExtractor.getData() and RedfinSalesDataExtractor.getData(). Each of these classes could have an interface called DataExtractor that defines that there must be a getData() method.

The class would then be something like 
```
class RedfinDataExtractor:
      driver = SeleniumWebDriverBuilder.build()
      listing_data_extractor = RedfinListingDataExtractor(driver)
      sales_data_extractor = RedfinSalesDataExtractor(driver)

     def get_listing_data():
         return listing_data_extractor.getData()

    def get_sales_data():
       return sales_data_extractor.getData()
```

Also, each page or element that you navigate to could be represented as objects and store the XPaths of the elements you'd like to interact with as constants. This way you'd have code that could be like 
```
filterPanel = FilterPanel(driver)
filterPanel.selectHomeTypes(home_types_list)
filterPanel.setFilterToOnlyActiveListings()
filterPanel.executeFilteringSearchAndClosePanel()
```

Reply:
Wow, you have become a fantastic software engineer. Reading this shows me how much you live and breath OOP and how much I, well, don't.

I'll likely be chatting with you more about some of these comments for some clarification, but this is all fantastic feedback. Thank you so much!